### ISSUE TITLE
openssl: error: target already defined

### ISSUE THREAD

**Issue by BigNoseSec** [Linuxbrew/homebrew-core#350]
Please follow the general troubleshooting steps first:

+ [x] Ran brew update and retried your prior step?
+ [x] Ran brew doctor, fixed as many issues as possible and retried your prior step?
+ [x] If you're seeing permission errors tried running sudo chown -R $(whoami) $(brew --prefix)?

Bug reports:

Having trouble installing cmake and qt. perl installed fine.

Using Archlinux, and other information needed can be provided. Anyone have any insight?

I tried sudo chown -R $(whoami) $(brew --prefix) and i just got the cowardly error. (as expected, so i reversed it back to root ownership.)

```
> sudo brew install cmake
> ==> Installing dependencies for cmake: openssl, bzip2, python, curl, libi
> ==> Installing cmake dependency: openssl
> ==> Downloading https://www.openssl.org/source/openssl-1.0.2h.tar.gz
> Already downloaded: /root/.cache/Homebrew/openssl-1.0.2h.tar.gz
> ==> perl ./Configure --prefix=/opt/linuxbrew/Cellar/openssl/1.0.2h --openssldir=
> Last 15 lines from /root/.cache/Homebrew/Logs/openssl/01.perl:
> 2016-05-25 15:13:17 -0400
> 
> perl
> ./Configure
> --prefix=/opt/linuxbrew/Cellar/openssl/1.0.2h
> --openssldir=/usr/local/etc/openssl
> no-ssl2
> zlib-dynamic
> shared
> enable-cms
>  -Os -w -pipe -march=native -Wl,--dynamic-linker=/usr/local/lib/ld.so -Wl,-rpath,/usr/local/lib
> linux-x86_64
> 
> target already defined -  -Os -w -pipe -march=native -Wl,--dynamic-linker=/usr/local/lib/ld.so -Wl,-rpath,/usr/local/lib (offending arg: linux-x86_64)
> 
> READ THIS: https://github.com/Linuxbrew/brew/blob/master/share/doc/homebrew/Troubleshooting.md#troubleshooting
> If reporting this issue please do so at (not Homebrew/brew):
>   https://github.com/Linuxbrew/homebrew-core/issues
```

**Comment by BigNoseSec** [comment]
Just updated the problem, dont think its permissions anymore.

**Comment by BigNoseSec** [comment]
So i tried to install openssl by itself and got roughly the same output:

`sudo brew install openssl                                                                 1 
==> Downloading https://www.openssl.org/source/openssl-1.0.2h.tar.gz
Already downloaded: /root/.cache/Homebrew/openssl-1.0.2h.tar.gz
==> perl ./Configure --prefix=/opt/linuxbrew/Cellar/openssl/1.0.2h --openssldir=/usr/local/etc/openssl no
Last 15 lines from /root/.cache/Homebrew/Logs/openssl/01.perl:
2016-05-25 15:50:34 -0400

perl
./Configure
--prefix=/opt/linuxbrew/Cellar/openssl/1.0.2h
--openssldir=/usr/local/etc/openssl
no-ssl2
zlib-dynamic
shared
enable-cms
 -Os -w -pipe -march=native -Wl,--dynamic-linker=/usr/local/lib/ld.so -Wl,-rpath,/usr/local/lib
linux-x86_64

target already defined -  -Os -w -pipe -march=native -Wl,--dynamic-linker=/usr/local/lib/ld.so -Wl,-rpath,/usr/local/lib (offending arg: linux-x86_64)

READ THIS: https://github.com/Linuxbrew/brew/blob/master/share/doc/homebrew/Troubleshooting.md#troubleshooting
If reporting this issue please do so at (not Homebrew/brew):
  https://github.com/Linuxbrew/homebrew-core/issues
`

**Comment by BigNoseSec** [comment]
After some time working on it I have begun to cry help pls.

**Comment by DoomHammer** [comment]
@sjackman is the expert here, but I think `brew` should not be run as root. Why are you using sudo in the first place?

**Comment by sjackman** [comment]
Yes, Piotr is correct. I would not suggest using `sudo` with Linuxbrew. The default is to install it in your home directory at `~/.linuxbrew`. I would suggest removing your existing installation of Linuxbrew, reinstall Linuxbrew without `sudo`, then `brew install openssl`. If it still fails, please report `brew gist-logs openssl`.

**Comment by sjackman** [comment]
If you create a user named `linuxbrew` and install to `/home/linuxbrew/.linuxbrew` you can use the precompiled binary bottle for `openssl`.

**Comment by cjprybol** [comment]
I'm running into this issue as well. I had been getting around it successfully with `brew install --force-bottle openssl`, although it seems that the bottle for openssl is no longer available. Doing a little digging, I found [this issue on the linuxbrew core repo](https://github.com/Linuxbrew/homebrew-core/issues/505), so I think I just got caught between a big bottle update.

I'm unable to make a new user and am stuck in root due to an issue that seems to be problematic for many container users.
```
Singularity.test.img> # adduser --disabled-password demo
Adding user `demo' ...
Adding new group `demo' (1002) ...
Adding new user `demo' (1002) with group `demo' ...
Creating home directory `/home/demo' ...
Copying files from `/etc/skel' ...
chfn: PAM: System error
adduser: `/usr/bin/chfn demo' returned error code 1. Exiting.
Singularity.test.img> # useradd another_demo
Singularity.test.img> # su another_demo
su: System error
```

I'm actually using a Singularity container, rather than a Docker container, but a quick google search on either error gets very similar issues for Docker users
https://github.com/docker/docker/issues/6345
http://stackoverflow.com/questions/25193161/chfn-pam-system-error-intermittently-in-docker-hub-builds
https://github.com/docker/docker/issues/5899
https://github.com/docker/docker/issues/5704
https://github.com/docker/docker/issues/11015

Any ideas? I'm happy to just wait for the new bottles to become available if that will be the quickest and easiest solution, but I'm not sure what the timeline for that would be. Are there any old bottle archives that would allow me to grab the previous bottle as a temporary fix while the new bottles are generated?

Thanks!

**Comment by rwhogg** [comment]
> Doing a little digging, I found this issue on the linuxbrew core repo, so I think I just got caught between a big bottle update.

Shaun merged homebrew/homebrew-core into linuxbrew/homebrew-core earlier today. We have to rebuild a lot of bottles as a result. The `openssl` bottle is [in progress](https://github.com/Linuxbrew/homebrew-core/pull/510) and high priority, so I wouldn't expect it to be long before it's available.

You can try `git checkout 6b749eddb1af3f8032e4a6276b5de9f67333f113`, reinstall `openssl`, and then `git checkout master` again to use the old bottle. Normally I would not recommend this since `openssl` has an unstable ABI, but in this case it should be harmless since the version number did not change (nor did any compiler flags).

**Comment by sjackman** [comment]
Here is the archive of old bottles: http://linuxbrew.bintray.com/bottles/
The `openssl` bottle has been updated!

**Comment by cjprybol** [comment]
Thank you to you both! For the quick turn around, and for developing Linuxbrew in general. It's really spectacular work.

I'll keep my eye on the checklist and use the archives in the meantime!

**Comment by sjackman** [comment]
Happy to help!

**Comment by paulmer** [comment]
I've run into this problem as well.  The issue is in line 51 of the formula.  If CPPFLAGS is empty, the result of joining cppflags, cflags and ldflags has a leading space which is causing the configure step to get confused.  Add ".strip" after the join to fix the problem.  I'll create a patch when I have a chance if no one else gets to it before me.

**Comment by paulmer** [comment]
Sorry everyone, I should have been clearer, the problem is in the openssl.rb rule.

**Comment by sjackman** [comment]
`CPPFLAGS` should never be empty. Its value should be `-isystem#{HOMEBREW_PREFIX}/include`. Unless, from what I can tell, Linuxbrew is installed in `/usr/local` (which is not recommended). Any chance that's the case here, that Linuxbrew is installed in `/usr/local`? If not, why is `CPPFLAGS` empty?

**Comment by sjackman** [comment]
Thanks for the troubleshooting, Paul! This error was a total mystery to me.

**Comment by cjprybol** [comment]
I had been installing to `/usr/local`, and this error goes away for me when I clone linuxbrew into it's own `.linuxbrew` directory elsewhere on the system. I can now install from source as well as from the bottle

**Comment by sjackman** [comment]
Glad to hear it, Cameron. I haven't tried installing in `/usr/local`. Except for this particular issue, has it otherwise worked well for you?

**Comment by cjprybol** [comment]
Yes, I haven't noticed any other problems at all. I've used the standard installation `ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Linuxbrew/install/master/install)"` and the `rm -r /usr/local && git clone https://github.com/Linuxbrew/brew.git /usr/local` installation both with success, alternating between the two depending on whether I'm a user (university clusters) or an admin (setting up containers). I gave the `/usr/local` install a try after seeing [this post](https://github.com/Linuxbrew/linuxbrew/issues/312#issuecomment-82004055), and everything that I've used thus far (primarily homebrew/science recipes and their dependencies) has worked using both setups!

**Comment by sjackman** [comment]
I could see potential problems if Linuxbrew installs its `glibc` into `/usr/local`. That could break your host system if its `glibc` version is different than Linuxbrew, currently 2.19 and same as Ubuntu 14 trusty. Glad to hear that it works for you!

**Comment by cjprybol** [comment]
Good to know, thank you for explaining Shaun! I'll switch to exclusive use of a separate `.linuxbrew` folder for both setups. I really had no better reason for using `/usr/local` except that it seemed empty and was already in the `$PATH`, neither of which are good enough reasons to risk a confusing breakage in the future

**Comment by sjackman** [comment]
Fixed by #541 thanks to @paulmer
